### DISCUSSION TITLE
Download from filer or s3 failed with: missing jwt

### DISCUSSION THREAD

**Discussion by n9** [seaweedfs/seaweedfs#4708] [Q&A]
When using JWT upload works, but download does not.

Log from volume:

```
I0726 14:14:35.865018 volume_server_handlers.go:165 missing jwt from 10.240.1.81:59900
I0726 14:14:35.865106 common.go:113 error JSON response status 401: wrong jwt
I0726 14:14:35.865139 common.go:70 response method:GET URL:/132,e1132f771a?readDeleted=true with httpStatus:401 and JSON:{"error":"wrong jwt"}
```

Log from filer:

```
E0726 14:14:35.865231 filer_server_handlers_read.go:258 failed to stream content /buckets/my-bucket/foo/bar2: read chunk: http://volume-0.volume.seaweed:8080/132,e1132f771a?readDeleted=true: 401 Unauthorized
E0726 14:14:35.865259 common.go:288 processRangeRequest: read chunk: http://volume-0.volume.seaweed:8080/132,e1132f771a?readDeleted=true: 401 Unauthorized
```

Tested both from the filer UI and by S3 client.

(Version: 3.55)

----

It seems to me that filer uses `ReadUrlAsStream` / `readEncryptedUrl` to read data from volume, it does not set JWT: https://github.com/seaweedfs/seaweedfs/blob/c09100f928466116fa648474417e30540026d8a4/weed/util/http_util.go#L290-L371 

But upload sets the JWT here:
https://github.com/seaweedfs/seaweedfs/blob/c09100f928466116fa648474417e30540026d8a4/weed/operation/upload_content.go#L319-L321

The JWT for upload is received here: https://github.com/seaweedfs/seaweedfs/blob/c09100f928466116fa648474417e30540026d8a4/weed/server/filer_server_handlers_write.go#L46

Should `LookupJwt` be used to get JWT for download?
https://github.com/seaweedfs/seaweedfs/blob/c09100f928466116fa648474417e30540026d8a4/weed/operation/assign_file_id.go#L106-L128

Or is it JWT only for write / delete? The only usage I have found in benchmark to authorize delete:
https://github.com/seaweedfs/seaweedfs/blob/c09100f928466116fa648474417e30540026d8a4/weed/command/benchmark.go#L213

Master uses the SigningKey to generate this JWT, not the ReadSigningKey: https://github.com/seaweedfs/seaweedfs/blob/c09100f928466116fa648474417e30540026d8a4/weed/server/master_grpc_server_volume.go#L102

But volume uses the ReadSigningKey for HEAD/GET:
https://github.com/seaweedfs/seaweedfs/blob/c09100f928466116fa648474417e30540026d8a4/weed/server/volume_server_handlers.go#L149-L161
https://github.com/seaweedfs/seaweedfs/blob/c09100f928466116fa648474417e30540026d8a4/weed/server/volume_server_handlers_read.go#L35

Is there a missing support for ReadSigningKey in the LookupVolume method?

Or could the JWT be computed in the filer instead of master?

In case that volume location is found in the `vidMap`, `LookupVolume` is not called:
https://github.com/seaweedfs/seaweedfs/blob/c09100f928466116fa648474417e30540026d8a4/weed/wdclient/masterclient.go#L59-L65

There would be an extra roundtrip to get JWT in that case?

----

UPDATE: This seems already reported here: https://github.com/seaweedfs/seaweedfs/issues/3352

And it is documented here: https://github.com/seaweedfs/seaweedfs/wiki/Security-Overview#jwt-for-read-access-control

**Comment by chrislusf** [comment]
https://github.com/seaweedfs/seaweedfs/wiki/Security-Overview 
filer has `jwt.filer_signing.key` and `jwt.filer_signing.read.key` in `security.toml`, right?

**Reply by n9** [reply]
Yes. But `jwt.filer_signing.key` and `jwt.filer_signing.read.key` are used to authorize a client of the filer. The problem is that in case of download, the filer request to volume is not authorized. (Volume says: "error JSON response status 401: wrong jwt".)
